ext/ddtrace.c

```diff
@@ -146,6 +147,100 @@ static void ddtrace_sort_modules(void *base, size_t count, size_t siz, compare_f
+// See also https://github.com/php/php-src/commit/f7c3f6e7e25471da9cfb2ba082a77cc3c85bc6ed
+static void dd_patched_zend_call_known_function(
+    zend_function *fn, zend_object *object, zend_class_entry *called_scope, zval *retval_ptr,
+    uint32_t param_count, zval *params, HashTable *named_params)
+{
+    zval retval;
+    zend_fcall_info fci;
+    zend_fcall_info_cache fcic;
+
+    // If current_execute_data is on the stack, move it to the VM stack
+    zend_execute_data *execute_data = EG(current_execute_data);
+    if ((uintptr_t)&retval > (uintptr_t)EX(func) && (uintptr_t)&retval - 0xfffff < (uintptr_t)EX(func)) {
+        zend_execute_data *call = zend_vm_stack_push_call_frame_ex(
+                ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_execute_data), sizeof(zval)) +
+                ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_op), sizeof(zval)) +
+                ZEND_MM_ALIGNED_SIZE_EX(sizeof(zend_function), sizeof(zval)),
+                0, EX(func), 0, NULL);
+
+        memcpy(call, execute_data, sizeof(zend_execute_data));
+        zend_op *opline = (zend_op *)(call + 1);
+        memcpy(opline, EX(opline), sizeof(zend_op));
+        zend_function *func = (zend_function *)(opline + 1);
+        func->common.fn_flags |= ZEND_ACC_CALL_VIA_TRAMPOLINE; // See https://github.com/php/php-src/commit/2f6a06ccb0ef78e6122bb9e67f9b8b1ad07776e1
+        memcpy((zend_op *)(call + 1) + 1, EX(func), sizeof(zend_function));
+
+        call->opline = opline;
+        call->func = func;
+
+        EG(current_execute_data) = call;
+    }
+
+    // here follows the original implementation of zend_call_known_function
+
+    fci.size = sizeof(fci);
+    fci.object = object;
+    fci.retval = retval_ptr ? retval_ptr : &retval;
+    fci.param_count = param_count;
+    fci.params = params;
+    fci.named_params = named_params;
+    ZVAL_UNDEF(&fci.function_name); /* Unused */
+
+    fcic.function_handler = fn;
+    fcic.object = object;
+    fcic.called_scope = called_scope;
+
+    zend_result result = zend_call_function(&fci, &fcic);
+    if (UNEXPECTED(result == FAILURE)) {
+        if (!EG(exception)) {
+            zend_error_noreturn(E_CORE_ERROR, "Couldn't execute method %s%s%s",
+                fn->common.scope ? ZSTR_VAL(fn->common.scope->name) : "",
+                fn->common.scope ? "::" : "", ZSTR_VAL(fn->common.function_name));
+        }
+    }
+
+    if (!retval_ptr) {
+        zval_ptr_dtor(&retval);
+    }
+}
+
+// We need to hijack zend_call_known_function as that's what's being called by call_attribute_constructor, and call_attribute_constructor itself is not exported.
+static void dd_patch_zend_call_known_function(void) {
+    size_t page_size = sysconf(_SC_PAGESIZE);
+    void *page = (void *)(~(page_size - 1) & (uintptr_t)zend_call_known_function);
+    if ((((uintptr_t)zend_call_known_function + 20) & page_size) < 20) {
+        page_size <<= 1; // if overlapping pages, use two
+    }
```

Comment:
The size of the assembly is 20.

Reply:
Many compiler options can influence assembly, so this patch makes me nervous anyway. How did you find the size of the function? `nm` or something? I wonder how hard it would be to find this size at runtime...

Reply:
No, it's the size of the trampoline assembl, which is hardcoded below. sizeof(absolute_jump_instrs) + sizeof(void*) for aarch64. And for the x64 it's just sizeof(absolute_jump_instrs).

Reply:
This code just takes care of the fact that we sometimes may need to write the trampoline across two pages, i.e. if the address of the func ends in 0xfff0